crates/pop-contracts/src/up.rs

```diff
@@ -93,9 +102,150 @@ pub async fn set_up_upload(
+	call_data.encode_call_data_to(&client.metadata(), &mut encoded_data)?;
+	Ok(encoded_data)
+}
+
+pub async fn get_instantiate_payload(
+	instantiate_exec: InstantiateExec<DefaultConfig, DefaultEnvironment, Keypair>,
+	gas_limit: Weight,
+) -> anyhow::Result<Vec<u8>> {
+	let storage_deposit_limit: Option<u128> = None;
+	let mut encoded_data = Vec::<u8>::new();
+	match instantiate_exec.args().code() {
+		contract_extrinsics::Code::Upload(code) => InstantiateWithCode::new(
+			instantiate_exec.args().value(),
+			gas_limit,
+			storage_deposit_limit,
+			code.clone(),
+			instantiate_exec.args().data().into(),
+			instantiate_exec.args().salt().into(),
+		)
+		.build()
+		.encode_call_data_to(&instantiate_exec.client().metadata(), &mut encoded_data),
+		contract_extrinsics::Code::Existing(hash) => Instantiate::new(
+			instantiate_exec.args().value(),
+			gas_limit,
+			storage_deposit_limit,
+			hash.clone(),
+			instantiate_exec.args().data().into(),
+			instantiate_exec.args().salt().into(),
+		)
+		.build()
+		.encode_call_data_to(&instantiate_exec.client().metadata(), &mut encoded_data),
+	}?;
+
+	Ok(encoded_data)
+}
+
+pub async fn get_contract_code(
+	path: Option<&PathBuf>,
+) -> anyhow::Result<contract_extrinsics::WasmCode> {
+	let manifest_path = get_manifest_path(path.map(|p| p as &Path))?;
+
+	// signer does not matter for this
+	let signer = create_signer("//Alice")?;
+	let extrinsic_opts =
+		ExtrinsicOptsBuilder::<DefaultConfig, DefaultEnvironment, Keypair>::new(signer)
+			.manifest_path(Some(manifest_path))
+			.done();
+	let artifacts = extrinsic_opts.contract_artifacts()?;
+
+	let artifacts_path = artifacts.artifact_path().to_path_buf();
+	let code = artifacts.code.ok_or_else(|| {
+		Error::UploadContractError(format!(
+			"Contract code not found from artifact file {}",
+			artifacts_path.display()
+		))
+	})?;
+	Ok(code)
+}
+
+pub async fn upload_contract_signed(
+	url: &str,
+	payload: String,
+) -> anyhow::Result<UploadResult<SubstrateConfig>> {
+	let events = submit_signed_payload(url, payload).await?;
+
+	let code_stored = events.find_first::<CodeStored<subxt::config::substrate::H256>>()?;
+
+	Ok(UploadResult { code_stored, events })
+}
+
+pub async fn instantiate_contract_signed(
+	url: &str,
+	payload: String,
+) -> anyhow::Result<InstantiateExecResult<SubstrateConfig>> {
+	let events = submit_signed_payload(url, payload).await?;
+
+	// The CodeStored event is only raised if the contract has not already been
+	// uploaded.
+	let code_hash = events
+		.find_first::<CodeStored<subxt::config::substrate::H256>>()?
+		.map(|code_stored| code_stored.code_hash);
+
+	let instantiated = events
+		.find_first::<ContractInstantiated<subxt::config::substrate::AccountId32>>()?
+		.ok_or_else(|| {
+			Error::InstantiateContractError("Failed to find Instantiated event".to_string())
+		})?;
+
+	Ok(InstantiateExecResult { events, code_hash, contract_address: instantiated.contract })
+}
+
+pub async fn submit_signed_payload(
```

Comment:
There’s potential duplicate code here, but we can refactor it once everything is finalized. Moving this into `pop_common`: https://github.com/r0gue-io/pop-cli/blob/main/crates/pop-parachains/src/call/mod.rs#L51, would allow us to reuse it. The difference is returning events, which is a work in progress here: https://github.com/r0gue-io/pop-cli/pull/372